### PR TITLE
refactor(assumptions): migrate common predicates to new design 

### PR DESCRIPTION
**Pull request by JSS95**
#### References to other Issues or PRs

Succeeds #20656

#### Brief description of what is fixed or changed
This PR migrates the part of old-design predicates to new design.
This PR concludes migrating the predicates. Old codes are removed and deprecated.

#### Release Notes

- assumptions
    - `AskHandler()`, `register_handler()` and `remove_handler()` are deprecated. Handler now must be multipledispatch instance.

### CODE REVIEW

File: sympy/assumptions/ask.py [hunk 1]
```diff
@@ -463,22 +437,17 @@ def ask_full_inference(proposition, assumptions, known_facts_cnf):
...
-        ...     @staticmethod
-        ...     def Integer(expr, assumptions):
-        ...         from sympy import log
-        ...         return ask(Q.integer(log(expr + 1, 2)))
-        >>> register_handler('mersenne', MersenneHandler)
-        >>> ask(Q.mersenne(7))
-        True
+    class inheriting from AskHandler.
+
+    .. note:: Deprecated in Sympy 1.8.
```
Comment (oscarbenjamin) [inline sympy/assumptions/ask.py:442]: There is `.. deprecated`:
https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-deprecated

File: sympy/assumptions/ask.py [hunk 2]
```diff
@@ -463,22 +437,17 @@ def ask_full_inference(proposition, assumptions, known_facts_cnf):
...
 
     """
-    # Will be deprecated
+    SymPyDeprecationWarning(
+        feature="register_handler() function",
+        useinstead="multipledispatch handler of Predicate",
+        issue=20656,
```
Comment (oscarbenjamin) [inline sympy/assumptions/ask.py:448]: The issue number should refer to an issue that is for the future removal of the deprecated feature. That issue should be written specifically to explain:

1. What is deprecated
2. What the new version is
3. How to change code that uses the deprecated API
4. When the deprecated codepaths will be removed completely.

The issue should have the "deprecation removal" label.

The wording of the linked issue should be written to make sense to a user who sees the warning. Bear in mind that the user might not have any idea which part of their code (or any other code they are using) has triggered the warning so there needs to be a clear explanation.

File: sympy/assumptions/tests/test_query.py [hunk 3]
```diff
@@ -2026,21 +2024,6 @@ def test_key_extensibility():
     # make sure the key is not defined
     raises(AttributeError, lambda: ask(Q.my_key(x)))
 
-    # Old handler system
-    class MyAskHandler(AskHandler):
-        @staticmethod
```
Comment (oscarbenjamin) [inline sympy/assumptions/tests/test_query.py:2031]: We should keep the tests for the old handler system while it is deprecated. You can use `warns_deprecated_sympy` to test that it still works and also gives a warning.

**Comment by JSS95** [comment]
Thanks. I made a deprecation issue #20837.

**Comment by JSS95** [comment]
Since every review is resolved, I'm merging this.

File: sympy/assumptions/tests/test_query.py [hunk 4]
```diff
@@ -2027,18 +2026,19 @@ def test_key_extensibility():
...
-            return True
-    try:
-        register_handler('my_key', MyAskHandler)
-        assert ask(Q.my_key(x)) is True
-        assert ask(Q.my_key(x + 1)) is None
-    finally:
-        remove_handler('my_key', MyAskHandler)
-        del Q.my_key
-    raises(AttributeError, lambda: ask(Q.my_key(x)))
+    with warns_deprecated_sympy():
```
Comment (oscarbenjamin) [inline sympy/assumptions/tests/test_query.py:2029]: This context manager should be used specifically around the line that emits the warning.

It only tests that the warning is emitted at least once so if there are multiple lines that warn then it needs to be used multiple times like:
```
with warns():
    line1
with warns():
    line2
```

**Comment by oscarbenjamin** [comment]
@JSS95 can you please wait for a positive review before merging?

**Comment by JSS95** [comment]
I'm sorry. I will.

**Pull request merged**